designs/use-selectors-at-cache.md

```diff
@@ -0,0 +1,125 @@
+Filter cache ListWatch using selectors
+===================
+## Motivation
+
+Controller-Runtime controllers use a cache to subscribe to events from
+Kubernetes objects and to read those objects more efficiently by avoiding
+to call out to the API. This cache is backed by Kubernetes informers.
+
+The only way to filter this cache is by namespace and resource type.
+In cases where a controller is only interested in a small subset of objects
+(for example all pods on a node), this might end up not being efficient enough.
+
+Requests to a client backed by a filtered cache for objects that do not match
+the filter will never return anything, so we need to make sure that we properly
+warn users to only use this when they are sure they know what they are doing.
+
+This proposal sidesteps the issue of "How to we plug this into the cache-backed
+client so that users get feedback when they request something that is
+not matching the caches filter" by only implementing the filter logic in the
+cache package. This allows advanced users to combine a filtered cache with the
+already existing `NewCacheFunc` option in the manager and cluster package,
+while simultaneously hiding it from newer users that might not be aware of the
+implications and the associated foot-shoot potential.
+
+The only alternative today to get a filtered cache with controller-runtime is
+to build it out-of tree. Because such a cache would mostly copy the existing
+cache and add just some options, this is not great for consumers.
+
+This proposal is related to the following issue [2]
+
+## Proposal
+
+Add a new selector code at `pkg/cache/internal/selector.go` with common structs
+and helpers
+```golang
+package internal
+
+...
+
+// SelectorsByObject associate a client.Object to a field/label selector
+type SelectorsByObject map[client.Object]Selector
```

Comment:
This should be restricted to `runtime.Object` so users don't get the idea that you can filter by object name/namespace
```suggestion
type SelectorsByObject map[runtime.Object]Selector
```

Reply:
Done

Reply:
@estroz why? It is a map, the values can not access the keys. Making this `runtime.Object` means we have to deal with stuff like ppl putting List types in there

Reply:
@alvaroaleman don't we have the same problem with client.Object since it's embedding runtime.Object ?

Reply:
@alvaroaleman it may be confusing to users to have all these object fields exposed by the `metav1.Object`, since they may assume the key can actually apply its own selector values, ex. by `SetName()`. Instead it would be better to constrain the contract to just `GroupVersionKind()`.

There can be some internal validation step to make sure the object is not a list type.

Reply:
Using GVK is inconvenient. Using the smaller interface (runtime.Object) doesn't prevent ppl from putting in objects that fulfill a lager interface (client.Object). It does however keep them from putting something like List types in there and makes the compiler verify that rather than getting a runtime error. Also, I think the idea that ppl might think the map key has  any further relevance for filtering when the value is a `Selector` struct seems a bit far fetched to me.

We also have plenty of prior art where we use client.Object to only derive the type, e.G. in source.Kind or the clients delete method.

Reply:
You could similarly argue that someone would not try to map a selector to a list type because that's not the type they add to `For()`, `Owns()`, or `Watches()`.

However since there is prior art I am fine with `client.Object`. There can always be another selector map type added later for `runtime.Object` without breakage if we see bugs filed because of confusion. It's not worth worrying about too much now.

Reply:
Going back to `client.Object` as proposed originally.